oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticResultRowIterator.java

```diff
@@ -713,21 +503,60 @@ private QueryBuilder createQuery(String propertyName, Filter.PropertyRestriction
         throw new IllegalStateException("PropertyRestriction not handled " + pr + " for index " + defn);
     }
 
-    class ElasticsearchFacetProvider implements FulltextIndex.FacetProvider {
+    class ElasticFacetProvider implements FulltextIndex.FacetProvider {
         private ElasticFacets elasticFacets;
         private Map<String, List<FulltextIndex.Facet>> cachedResults = new HashMap<>();
-
-        ElasticsearchFacetProvider(ElasticFacets elasticFacets) {
-            this.elasticFacets = elasticFacets;
-        }
+        private AtomicBoolean isInitialized = new AtomicBoolean();
 
         @Override
         public List<FulltextIndex.Facet> getFacets(int numberOfFacets, String columnName) throws IOException {
-            String facetProp = FulltextIndex.parseFacetField(columnName);
-            if (cachedResults.get(facetProp) == null) {
-                cachedResults = elasticFacets.getFacets(indexNode.getDefinition(), numberOfFacets);
+            if (isInitiliazed()) {
+                String facetProp = FulltextIndex.parseFacetField(columnName);
+                if (cachedResults.get(facetProp) == null) {
+                    cachedResults = elasticFacets.getFacets(rowIteratorState.indexNode.getDefinition(), numberOfFacets);
+                }
+                return cachedResults.get(facetProp);
+            } else {
+                LOG.error("FacetProvider not initialized");
             }
-            return cachedResults.get(facetProp);
+            return Collections.emptyList();
+        }
+
+        public boolean isInitiliazed() {
+            return isInitialized.get();
+        }
+
+        public void initialize(ElasticFacets elasticFacets) {
+            isInitialized.set(true);
+            this.elasticFacets = elasticFacets;
         }
     }
+
+    class ElasticRowIteratorState {
+
+        final Deque<FulltextIndex.FulltextResultRow> queue = new ArrayDeque<>();
+        // TODO : find if ES can return dup docs - if so how to avoid
+        SearchHit lastDoc;
+        boolean noDocs = false;
+
+        final Filter filter;
+        final FulltextIndexPlanner.PlanResult planResult;
+        final QueryIndex.IndexPlan plan;
+        final ElasticIndexNode indexNode;
+        final ElasticResultRowIterator.RowInclusionPredicate rowInclusionPredicate;
+        final LMSEstimator estimator;
```

Comment:
We are exposing everything (which were earlier private) to other classes from this class. May be we can encapsulate better and expose only those things which are required. I understand we are using this as a state class but why not add methods as well to encapsulate things better (then we may rename the class :) ). 
For example we are doing -
`elasticRowIteratorState.estimator.update(elasticRowIteratorState.filter, docs.getHits().getTotalHits().value);`
We could add a method to this class -
`updateEstimates(long value)`

Likewise the static method `getPathRestrictionQuery` uses three of the fields from this class. Can we add that method to this class itself (as non static method)?

Reply:
Added getter for the fields.
According to me getPathRestrictionQuery should be in ElasticQueryUtil only. I think we can add this method in state class when we change its name.